ouroboros-consensus/docs/utxo-hd/report.tex

```diff
@@ -0,0 +1,741 @@
+
+The LMDB\footnote{http://www.lmdb.tech/doc/} implementation uses a key-value store on the disk and things get somewhat more complicated. The operations are achieved as follows:
+
+\begin{itemize}
+  \item \textbf{WRITE}: LMDB provides an API function for inserting/updating/deleting values.
+  \item \textbf{READ}: Holding onto a \emph{read} LMDB transaction provides a
+        consistent view of the database for as long as the transaction is kept
+        open.
+  \item \textbf{SNAPSHOT}: LMDB provides an API function to copy a whole
+        database onto a separate directory.
+\end{itemize}
+
+The data is essentially split in two tables, namely \emph{dbState} (which
+holds the sequence number explained above), and a table with a name
+dictated by the actual block type in use by the node.
+
+\subsubsection{LSM Trees}
+
+TBD
+
+\section{Interface with the Ledger}
+
+The Ledger layer has not been modified for UTxO-HD and therefore the ledger
+states that we provide to the ledger must be indistinguishable from those before
+UTxO-HD in what matters to the Ledger rules that will be executed. In order to
+understand why some actions are allowed, one has to interiorize the following
+\textbf{Key observations}:
+
+\begin{observation}
+  \emph{In order to succesfully apply a block, the UTxO set inside the ledger state
+  only needs to contain the entries requested by the transactions in the block.}
+\end{observation}
+
+This is also true when speaking about transaction application: only the inputs
+requested by the transaction are actually needed by the Ledger.
+
+\begin{observation}
+  \emph{In order to succesfully tick a block not through an era transition, the
+    UTxO set can be empty as the ledger rules don't look at it.}
+\end{observation}
+
+\begin{observation}
+  \emph{Ticking through an era transition might require some entries on the UTxO set
+  and can produce changes in them.}
+\end{observation}
+
+It is precisely by this last observation, that ticking has to be considered as
+producing tables with differences.
+
+The sequence of actions for ticking and applying a block $B$ at slot $s$ on top
+of a ledger state $L$ is then easily described by \textbf{Algorithm \ref{alg:tickandapply}}.
+
+\begin{algorithm}[h]
+  \caption{Tick and apply}
+  \label{alg:tickandapply}
+  \begin{algorithmic}[1]
+    \Procedure{TickAndApply}{$DB :: \texttt{LedgerDB}, B :: \texttt{Block}$}
+    \State $L_{E} \gets \texttt{getLedger}_{DB}(\texttt{Pred}(B))$
+    \State $\mathbb{L}^{s} \gets \texttt{TICK}(\mathbb{L}, s)$ \Comment{Note that $\mathbb{L} \eqsim L_{E}$}
+    \State $\texttt{Diffs}_{s} \gets \texttt{diff}(\mathbb{L}^{s}, \mathbb{L})$
+    \Statex
+    \State $V_{in} \gets \texttt{queryDisk}_{DB}(\texttt{keySets}(B)) \triangleleft \texttt{Diffs}_{DB}(\texttt{Pred}(B)) \triangleleft \texttt{Diffs}_{s}$ \Comment{$\triangleleft$ means \textbf{forward}}
+    \Statex
+    \State $\mathbb{L}^{s}_{[V_{in}]} \gets \texttt{stowLedgerTables}(L^{s}_{{V_{in}}})$
```

Comment:
The notation involving `[]` in the index are not explained; this could be done when `StowableLedgerTables` is introduced above.